### PR TITLE
feat: use Gravatar for displaying user's profile avatar

### PR DESCRIPTION
**Pull request by wellyshen**
## What does this PR do?

To implement #3172 

## How should this be tested?

- Log in using an email associated with a Gravatar avatar (uploaded from here: https://gravatar.com/profile/avatars)
- Navigate to the user's account settings page (`/settings/profile`), and you'll see the avatar already displayed :)

| Before | After |
| - | - |
| 截圖 2024-09-25 下午5 17 09 | 截圖 2024-09-25 下午5 21 08 |

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [x] First PR at Formbricks? [Please sign the CLA!](https://cla-assistant.io/formbricks/formbricks) Without it we wont be able to merge it 🙏

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Formbricks Docs if changes were necessary

## Summary by CodeRabbit

- **New Features**
	- Updated `ProfileAvatar` component to use email for user identification instead of user ID.
	- Enhanced image handling configuration to support images from Gravatar.
	- Streamlined data handling in the `EditProfileAvatarForm` for improved user experience.

- **Bug Fixes**
	- Removed unnecessary session prop from components, streamlining user data handling.

- **Documentation**
	- Updated prop signatures and component usage in relevant files to reflect changes in user identification methods.

### CODE REVIEW

File: packages/ui/Avatars/index.tsx
Comment (wellyshen) [inline packages/ui/Avatars/index.tsx:21]: Gravatar allows for a fallback default image using the d URL parameter. We can also set Boring Avatar as the default by utilizing [their API service](https://boringavatars.com/#api-service).

**Comment by jobenjada** [comment]
hey @wellyshen  - thanks for this! :)

Any insights on what that means in terms of data privacy? It looks like we'd be sharing the user email (PII) with Gravatar to find their image.

Since our focus is on privacy, this might be an issue 🤓

Cheers!

**Comment by wellyshen** [comment]
> Any insights on what that means in terms of data privacy? It looks like we'd be sharing the user email (PII) with Gravatar to find their image.

Hey @jobenjada,

Thank you for the review.

The email provided to Gravatar is hashed using SHA-256, which is a one-way cryptographic function. In theory, it’s impossible to reverse a hash created by SHA-256. Therefore, I believe user privacy is well-protected when using Gravatar to display default avatars.

File: packages/ui/components/Avatars/index.tsx
Comment (mattinannt) [inline packages/ui/components/Avatars/index.tsx:2]: Please use the `crypto` package already included in node.js instead of using an extra package:

```
function generateSHA256Hash(input) {
  return createHash('sha256').update(input).digest('hex');
}
```

File: packages/ui/components/Avatars/index.tsx
Comment (mattinannt) [inline packages/ui/components/Avatars/index.tsx:24]: Since you are using the next.js image component you need to whitelist the gravatar domain in the `apps/web/next.config.mjs` config :-) 
(`nextConfig.images.remotePatterns`)

File: packages/ui/components/Avatars/index.tsx
Comment (mattinannt) [inline packages/ui/components/Avatars/index.tsx:31]: Did you remove the `<Avatar />` component? I think it's still a good idea to return the boring-avatar if neither an image is uploaded by the user nor a gravatar is present since I think this is the most common case.

**Review by mattinannt — CHANGES_REQUESTED** [review]
Thanks for the PR 💪 But I found a few issues that should be solved before we can merge this :-)

File: packages/ui/components/Avatars/index.tsx
Comment (wellyshen) [reply packages/ui/components/Avatars/index.tsx:31]: Yes, because Gravatar will return a default image (e.g., d=retro), the <Avatar /> won't be needed.

**Comment by wellyshen** [comment]
> Thanks for the PR 💪 But I found a few issues that should be solved before we can merge this :-)

Thank you for the feedback, I will address them.

File: packages/ui/components/Avatars/index.tsx
Comment (wellyshen) [reply packages/ui/components/Avatars/index.tsx:25]: ✅

**Comment by wellyshen** [comment]
@mattinannt I have addressed the feedback, please take a look when you get a moment, thank you!

File: packages/ui/components/Avatars/index.tsx
Comment (wellyshen) [reply packages/ui/components/Avatars/index.tsx]: ✅

File: packages/ui/components/Avatars/index.tsx
Comment (mattinannt) [reply packages/ui/components/Avatars/index.tsx:21]: @wellyshen CodeRabbit is partially right. I think we should force the email to be lowercase first since that's what the docs say: https://docs.gravatar.com/api/avatars/hash/

(we likely already have the email addresses in our system already lowercase but makes sense to also have this here since there might be non-lowercase emails in the db)

File: packages/ui/components/Avatars/index.tsx
Comment (mattinannt) [reply packages/ui/components/Avatars/index.tsx:32]: I would prefer using our boring avatars as default since I'm not a fan of the default gravatar images ;-)

File: packages/ui/components/Avatars/index.tsx
Comment (mattinannt) [reply packages/ui/components/Avatars/index.tsx:2]: This can be an issue in the future.
Please add `import "server-only";` to the file to make sure it's only used on the server.

**Review by mattinannt — CHANGES_REQUESTED** [review]
@wellyshen Thanks for the fixes 💪 Reviewing this again I still have a few smaller issues but we are nearly good to go :-)

File: packages/ui/components/Avatars/index.tsx
Comment (wellyshen) [reply packages/ui/components/Avatars/index.tsx:32]: @mattinannt Gravatar will always return a default image if no avatar is found. Unfortunately, we cannot fall back to the boring-avatar React component. However, we can replace Gravatar's default image using the boring-avatar API, like this: https://www.gravatar.com/avatar/${hashedEmail}?d={CUSTOM_DOMAIN}?name={NAME}&variant={VARIANT}&size={SIZE}&colors={COLORS}&square.

Are you planning to use this service?

**Comment by wellyshen** [comment]
@mattinannt I have addressed the email format issue and also left two comments, please take a look, thank you!

**Comment by mattinannt** [comment]
@wellyshen Thanks a lot for making the changes 💪
Not having the default avatar is a bit of an issue for us. A lot of our self-hosting users are running this behind a firewall that blocks outside access so they also wouldn't be able to load the default gravatar image. We would either need a good fallback or even a way to enable/disable the gravatar integration and I think this would be a bit of an overkill for a nice-to-have feature like gravatars.
We are in a bit of a dilemma of rather adding a lot of complexity, or sacrificing avatars in self-hosting environments, or not being able to use gravatars. It seems like we can't achieve both.

Unfortunately because of this I would close the PR. We really appreciate the work you put into this and we are sorry that we cannot merge this at the moment. We can still provide you with oss.gg hackathon points for this PR if you like :-)

**Comment by wellyshen** [comment]
@mattinannt Thank you for the explanation. I really appreciate your code review :)

**Pull request closed without merging**
